### PR TITLE
Feature: Select audio device

### LINKED ISSUE #17: mic recording may default to webcam
Mori noticed this while testing, not the bug of the century but probably should make it configurable.

### PR DESCRIPTION
**Pull request by tdaugaard**
Implement a way to seelct input and output devices for recording via Settings. Fixes #17

### CODE REVIEW

**Comment by tdaugaard** [comment]
I'm unsure exactly how the device scenes work - does it actually capture audio from _all_ devices as it is now, or does it just select one device out of however many there is?

I created two extra options besides the actual audio devices: "All" and "None", but as I'm unsure what it does exactly, maybe only "None" + devices are required. Let me know, if you know :)

**Comment by aza547** [comment]
Looks great as usual. You should increase the height of the settingsWindowin main.ts so that it all fits. 

> I'm unsure exactly how the device scenes work - does it actually capture audio from _all_ devices as it is now, or does it just select one device out of however many there is?
> 


> I created two extra options besides the actual audio devices: "All" and "None", but as I'm unsure what it does exactly, maybe only "None" + devices are required. Let me know, if you know :)

Not sure of the top of my head how this works, I'll have a look when I get some time and get back to you.

**Comment by aza547** [comment]
I don't really understand why I can't see obsRecorder.reconfigure() here. Have you branched off an old version of main? I think you need to ensure that function works (think you will need to plumb some config through as args).

File: src/main/obsRecorder.ts [hunk 1]
```diff
@@ -158,44 +160,45 @@ const setupScene = (monitorIndex: number) => {
...
-  dummyDevice.release();
-  return devices;
-};
-
 /*
 * setupSources
 */
 const setupSources = (scene: any) => {
+  const cfg = new Store();
+  const selectedInputAudioDevice = (cfg.get('audio-input-device', 'all') as string);
```
Comment (aza547) [inline src/main/obsRecorder.ts:168]: The existing model is to keep track of config in main.ts and pass values through to the Recorder class as args. I think this makes reconfiguring easier.

Comment (tdaugaard) [reply src/main/obsRecorder.ts:168]: Honestly, I hadn't even noticed that `cfg.get()` was exposed via an IPC channel. 🤦‍♀️ I will fix, of course :)

Comment (tdaugaard) [reply src/main/obsRecorder.ts]: Fixed to be in line with the existing way of passing configuration to the `Recorder`.

**Comment by tdaugaard** [comment]
I have fixed the issue you addressed here. I was contemplating creating an object/type to contain and pass the settings to the recorder through, to avoid having to pass 4 arguments everywhere, but I held back and kept to the actual feature 😂

**Comment by aza547** [comment]
Looks good.

**Comment by aza547** [comment]
Have you tested? Particularly interested in the reconfigure function.

**Comment by aza547** [comment]
Had a quick go at testing. Didn't get any sound in the video. My config below for ref: 

image

Also some errors in react:

image

**Comment by tdaugaard** [comment]
> Had a quick go at testing. Didn't get any sound in the video. My config below for ref:

I was testing it quite a lot yesterday with no audio, only mic, no output device, other way around, different mic (BT headset) etc. and it worked fine exactly as intended.

I did not see the errors in the console - I tend to close the inspector tab because it takes up too much space when testing 😬 I'm fixing those now.

**Comment by tdaugaard** [comment]
> You should increase the height of the settingsWindow in main.ts so that it all fits. This page is probably due a revamp at some stage to be better setup.

Definitely, but it'll do for now no problem. I changed the size to 650x460 to allow two columns instead of having it become very tall. If you prefer a tall window, I'll make it tall =)

**Comment by aza547** [comment]
> > You should increase the height of the settingsWindow in main.ts so that it all fits. This page is probably due a revamp at some stage to be better setup.
> 
> Definitely, but it'll do for now no problem. I changed the size to 650x460 to allow two columns instead of having it become very tall. If you prefer a tall window, I'll make it tall =)

Wide is good too :)

**Comment by tdaugaard** [comment]
I was able to reproduce the problem you mentioned about no output audio being recorded. I'll investigate and test further.

**Comment by tdaugaard** [comment]
I had mixed up the input/output device selection _gravely_ which affected the feature in a way that made it never record any output audio. This has been fixed.

There's a small bug in the settings window where it won't show the selected option after selecting it, but it'll save it properly and reload it as intended when the settings window is opened once again.

I'll investigte this and fix it, but that should be the last thing.

**Comment by aza547** [comment]
Had a test, got sound which was nice!

Had a problem where I did a recording with sound, then changed settings to No input / No output.

Recorded again and still got sound.

**Comment by tdaugaard** [comment]
> Had a problem where I did a recording with sound, then changed settings to No input / No output.

I am able to reproduce this - am looking into it.

**Comment by tdaugaard** [comment]
If I select a device, record, then I get audio. Then select None, and I still get audio. Then select a device, THEN select None, then Update and now I get no sound. Interesting.

**Comment by tdaugaard** [comment]
It was due to the way audio tracks are added; it's like the old tracks aren't removed when `setupSources()` runs subsequently. 

Temporily fixed by using `source.muted` flag to mute the sources that we haven't selected, but we should find a way to clear the sources before configuring them again.

**Comment by tdaugaard** [comment]
After some more testing, it _is_ possible to simply delete existing sources first:

```typescript
for (let i = 0; i < 12; i++) {
  osn.Global.setOutputSource(i, null);
}
```

But this has the downside that we need to know how many sources there were, since `osn.Global` doesn't expose that, as far as I can tell.

I'll let this one stay as it is, because it works.

**Comment by aza547** [comment]
This is a really awesome feature. I didn't find any bugs in the latest test. Nice work.

File: src/main/util.ts [hunk 2]
```diff
@@ -492,13 +493,13 @@ const cutVideo = async (initialFile: string, finalDir: string, desiredDuration:
...
- * Gets string value from the config in a more reliable manner.
+ * Gets a path (string) value from the config in a more reliable manner.
  * @param cfg the config store
  * @param key the key
  * @returns the string config
  */
-const getPathConfigSafe = (cfg: any, key: string): string => {
-    return cfg.has(key) ? path.join(cfg.get(key), "/") : "";
+const getPathConfigSafe = (cfg: ElectronStore, key: string): string => {
+    return cfg.has(key) ? path.join(getStringConfigSafe(cfg, key), path.sep) : "";
```
Comment (aza547) [inline src/main/util.ts:502]: path.sep, nice!

File: src/main/operatingSystems.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@ const OS = {
     Mac: 'darwin',
   }
   
-  function byOS(handlers) {
+  function byOS(handlers: any) {
```
Comment (aza547) [inline src/main/operatingSystems.ts:8]: Pretty sure this is all useless code FYI, inherited from the obsRecorder copy paste.

Comment (tdaugaard) [reply src/main/operatingSystems.ts:8]: Oh, yeah it's there to basically support MacOS as well from what I can gather. If you don't plan on supporting that, we can remove it - if you want to keep the support in (for that part of the code anyway) then we should keep it.

**Pull request merged**